training/CHAPEL_TUTORIAL.md

```diff
@@ -0,0 +1,846 @@
+2
+6
+3
+1
+7
+5
+```
+As we can see this loop is not executing serially.
+Your output will likely be in a different order than the above and will differ between runs.
+
+<a id="ex8"></a>
+#### Example 8: may-parallel `forall` Expression
+`forall`s can also be used in expressions, for example:
+```Chapel
+var tens = [i in 1..10] i*10;
+writeln(tens);
+var negativeTens = [i in tens] -i;
+writeln(negativeTens);
+```
+```console
+10 20 30 40 50 60 70 80 90 100
+-10 -20 -30 -40 -50 -60 -70 -80 -90 -100
+```
+<a id="ex9"></a>
+#### Example 9: Factorial with may-parallel `forall` Expression and Reduction
+Applying a may-parallel `forall` expression and a reduction, our factorial function becomes:
+```Chapel
+proc factorial(n: int) {
+  return * reduce [i in 1..n] i;
+}
+writeln(factorial(5));
+```
+```console
+120
+```
+
+For a specific `n`, we  can do this in one line!
+```Chapel
+writeln(* reduce [i in 1..#5] i);
+```
+
+<a id="TIY1"></a>
+#### Try It Yourself 1: Perfect Squares <=25
+Problem:
+Generate and print an array of all perfect squares less than or equal to `25`
+
+Bonus points if you can do it in one line using a `forall` expression!
+<details>
+  <summary>Potential Solution</summary>
+
+```Chapel
+writeln([i in 0..5] i**2);
+```
+</details>
+Expected Output:
+
+```console
+0 1 4 9 16 25
+```
+
+<a id="ternary"></a>
+## Ternary
+A ternary statement is where a variable can have one of two possible values depending on whether a condition is True or False.
+
+The syntax for a ternary statement is:
+```chapel
+var x = if cond then val1 else val2;
+```
+This is equivalent to an `if/else`:
+```chapel
+var x: t;
+if cond {
+  x = val1;
+}
+else {
+  x = val2;
+}
+```
+<a id="ex10"></a>
+#### Example 10: Absolute Value Ternary
+Let's use a ternary to create an absolute value function:
+```chapel
+proc absoluteVal(n:int) {
+  return if n >= 0 then n else -n;
+}
+writeln(absoluteVal(-15));
+writeln(absoluteVal(7));
+```
+```console
+15
+7
+```
+<a id="ex11"></a>
+#### Example 11: Ternary and `forall` Expression
+We can combine ternary with other tools such as forall expressions.
+```chapel
+writeln([i in 0..#10] if i%2 == 0 then i+10 else -100);
+```
+```console
+10 -100 12 -100 14 -100 16 -100 18 -100
+```
+
+<a id="TIY2"></a>
+#### Try It Yourself 2: Array Absolute Value
+Problem:
+Write a `proc` using a ternary to take an `int array`, `A`, 
+and return an array where index `i` is the absolute value of `A[i]`
+
+Input: 
+`[-3, 7, 0, -4, 12]`
+
+<details>
+  <summary>Potential Solution</summary>
+
+```Chapel
+proc arrayAbsVal(A) {
+  return [a in A] if a >= 0 then a else -a;
+}
+```
+</details>
+Expected Output:
+
+```console
+3 7 0 4 12
+```
+
+<a id="introspection"></a>
+## Introspection
+Introspection is determining properties of a function argument at runtime.
+This is often used to determine the type and/or domain of a function argument. 
+Using introspection can avoid duplicating a `proc` for multiple types when none of the logic has changed.
```

Comment:
I would add a note about this making the function `generic` since we have had to debug multiple things due to generic function recently. I think a simple mention with a link to the Chapel documentation would be good.